Jenkinsfile

```diff
@@ -0,0 +1,92 @@
+dockerImage = 'soramitsu/kagome-dev:6'
+workerLabel = 'd3-build-agent'
+buildDir = "/tmp/build"
+
+
+def makeBuild(String clangFlags, Closure then) {
+  return {
+    try {
+      node(workerLabel) {
+        scmVars = checkout scm
+        sh(script: "git pull --recurse-submodules")
+        sh(script: "git submodule update --init --recursive")
+        withCredentials([
+          string(credentialsId: 'codecov-token', variable: 'codecov_token'),
+          usernamePassword(credentialsId: 'sorabot-github-user', passwordVariable: 'sorabot_password', usernameVariable: 'sorabot_username')
+        ]) {
+          docker.image(dockerImage).inside('--cap-add SYS_PTRACE') {
+
+            // define env to enable hunter cache
+            withEnv([
+              "CTEST_OUTPUT_ON_FAILURE=1",
+              "GITHUB_HUNTER_USERNAME=${sorabot_username}",
+              "GITHUB_HUNTER_TOKEN=${sorabot_password}",
+              "CODECOV_TOKEN=${codecov_token}"
+            ]) {
+              sh(script: "cmake . -B${buildDir} ${clangFlags}")
+              // continue execution
+              then()
+            } // end withEnv
+          } // end docker.image
+        } // end withCredentials
+      } // end node
+    } catch(Exception e) {
+      currentBuild.result = 'FAILURE'
+      error(e)
+    } // end catch
+  } // end return
+} // end makeBuild
+
+def makeToolchainBuild(String toolchain) {
+  return makeBuild("-DCMAKE_TOOLCHAIN_FILE=${toolchain}", {
```

Comment:
I would suggest to wrap functions which represent stages into `stage` stanza
```suggestion
  return stage('Toolchain Build') {
    makeBuild("-DCMAKE_TOOLCHAIN_FILE=${toolchain}", { ... }
  }
```
This makes Jenkins to nicely show each stage and its duration in the build. Right now there is a single stage:
![image](https://user-images.githubusercontent.com/8650477/61198339-372c9880-a6e2-11e9-8f01-655185ecfc93.png)